interfaces/cython/cantera/yaml2ck.py

```diff
@@ -0,0 +1,1047 @@
+                break_long_words=False,
+                break_on_hyphens=False,
+            )
+
+            species_txt = f"{species_names}"
+
+        text = ["SPECIES", f"{species_txt}", "END\n\n"]
+
+        return "\n".join(text)
+
+    text = []
+    text.append(build_elements(solution.element_names, solution.atomic_weights))
+    text.append(build_species(solution.species()))
+
+    return "\n".join(text)
+
+
+def build_thermodynamics_text(solution_species, separate_file=False):
+    """
+    Creates the thermodynamic definition text of all species
+
+    :param solution_species:
+        The species from the `solution` object being converted
+    :param separate_file:
+        A boolean flag to indicate if the file will be written separately or
+        in the mechanism file
+    """
+
+    def build_thermo_notes(solution_species):
+        """
+        Tries to infer if a note is within the NASA7 definition or is a note to be
+        placed as a comment.
+        """
+        comment = {}
+        note = {}
+        for species in solution_species:
+            if "note" in species.input_data["thermo"]:
+                note_raw = species.input_data["thermo"]["note"]
+            else:
+                note_raw = ""
+
+            # attempt to split note and comment
+            if len(note_raw.split("\n", 1)) == 1:
+                comment[species.name] = ""
+                comment_str = ""
+                note[species.name] = note_raw
+            else:
+                comment[species.name] = "!\n"
+                note[species.name], comment_str = note_raw.split("\n", 1)
+
+            if len(f"{species.name} {note[species.name]}") > 24:
+                comment_str += "\n" + note[species.name]
+                note[species.name] = ""
+
+            comment_str = comment_str.replace("\n", "\n! ")
+            comment[species.name] = f"{comment[species.name]}! {comment_str}"
+
+        return note, comment
+
+    def build_nasa7_text(solution_species):
+        # first line has species name, space for notes/date, elemental composition,
+        # phase, thermodynamic range temperatures (low, high, middle), and a "1"
+        # total length should be 80
+        #
+        # Ex:
+        # C6 linear biradi  T04/09C  6.   0.   0.   0.G   200.000  6000.000 1000.        1
+        #  1.06841281E+01 5.62944075E-03-2.13152905E-06 3.56133777E-10-2.18273469E-14    2
+        #  1.43741693E+05-2.87959136E+01 3.06949687E+00 3.71386246E-02-5.95698852E-05    3
+        #  5.15924485E-08-1.77143386E-11 1.45477274E+05 8.35844575E+00 1.47610437E+05    4
+
+        # get notes and comments for all species
+        note, comment = build_thermo_notes(solution_species)
+
+        # write data for each species in the Solution object
+        thermo_text = []
+        for species in solution_species:
+            composition_string = "".join(
+                [f"{s:2}{int(v):>3}" for s, v in species.composition.items()]
+            )
```

Comment:
I think we need to be more strict here about the format. According to the standard ([here](https://www3.nd.edu/~powers/ame.60636/chemkin2000.pdf) and [here](https://shepherd.caltech.edu/EDL/PublicResources/sdt/formats/chemkin.html) have the same format), this should be at most 4[A2I3]. More than 4 elements need to go after the temperatures, or after the line number.

Reply:
I am unfamiliar with the format of >4 element species. I just threw an exception in for now.